src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs

```diff
@@ -2055,12 +2055,14 @@ internal void ReadStateFile(FileExists fileExists)
         /// </summary>
         internal void WriteStateFile()
         {
-            if (!String.IsNullOrEmpty(AssemblyInformationCacheOutputPath))
+            if (!string.IsNullOrEmpty(AssemblyInformationCacheOutputPath))
             {
                 _cache.SerializePrecomputedCache(AssemblyInformationCacheOutputPath, Log);
             }
-            else if (!String.IsNullOrEmpty(_stateFile) && _cache.IsDirty)
+            else if (!string.IsNullOrEmpty(_stateFile) && (_cache.IsDirty || _cache.instanceLocalOutgoingFileStateCache.Count < _cache.instanceLocalFileStateCache.Count))
             {
+                // Either the cache is dirty (we added or updated an item) or the number of items actually used is less than what
+                // we got by reading the state file prior to execution. Serialize the cache into the state file.
```

Comment:
Not sure I understand the "the cache had an item in it that we didn't directly need but everything else was up to date, so remove it from the serialized cache" behavior. Is there a cost to leaving extra entries there?

Reply:
Rereading the summary--is this mostly for cleanup from the previous "throw the kitchen sink in there" case?

Reply:
> Is there a cost to leaving extra entries there?

The cost is deserializing the extra entries on every incremental build, until the developer does a clean build. There are these pathological cases like leaving the the entire pre-cache there or transitioning from the old behavior of serializing entries for SDK assemblies where it makes a big difference. But I believe it's better to be deterministic regardless, especially now that there are scenarios like DevBox where we may be distributing pre-built `obj`.

Reply:
The only downside I can think of is something like "switching back and forth between branches that add/remove a reference", but that's not a critical scenario.